### PR TITLE
Migration to cats-effect 3

### PR DESCRIPTION
**Pull request by tpetillot**

### CODE REVIEW

**Comment by tpetillot** [comment]
🚫 compilation failed in scala 2.13...

File: build.sbt
Comment (vkostyukov) [inline build.sbt:19]: We used to pin CE version to 2.x in scala-steward's config. Should we unpin now?

File: benchmarks/src/main/scala/io/finch/benchmarks.scala
Comment (vkostyukov) [inline benchmarks/src/main/scala/io/finch/benchmarks.scala:26]: Isn't there an instance of a Dispatcher for IO already?

File: core/src/main/scala/io/finch/Bootstrap.scala
Comment (vkostyukov) [inline core/src/main/scala/io/finch/Bootstrap.scala:85]: I have very limited knowledge of CE3, but from what I'm reading we don't need to ask for Dispatcher[F] if we already have Async[F]. Looks like we can just summon a dispatcher: https://typelevel.org/cats-effect/docs/std/dispatcher

File: core/src/main/scala/io/finch/Endpoint.scala
Comment (vkostyukov) [inline core/src/main/scala/io/finch/Endpoint.scala:303]: Same question about Dispatcher. Maybe it's enough just to have Async?

File: core/src/main/scala/io/finch/Endpoint.scala
Comment (vkostyukov) [inline core/src/main/scala/io/finch/Endpoint.scala:696]: Should we take implicit EC?

File: core/src/main/scala/io/finch/EndpointResult.scala
Comment (vkostyukov) [inline core/src/main/scala/io/finch/EndpointResult.scala:43]: I think we should ask for Async[F] and summon Dispatcher instead.

Something along the lines:

```
Dispatcher[F].use { dispatcher =>
case EndpointResult.Matched(_, _, out) =>
      try Some(Right(dispatcher.unsafeRunTimed(out, d)))
      catch {
        case _: TimeoutException => Some(Left(new TimeoutException(s"Output wasn't returned in time: $d")))
        case e: Throwable        => Some(Left(e))
      }
}
```

File: core/src/main/scala/io/finch/Input.scala
Comment (vkostyukov) [inline core/src/main/scala/io/finch/Input.scala:106]: Same question here:

I think we should keep asking for Async type-class and derive Dispatcher out of it later.

File: core/src/test/scala/io/finch/EndToEndSpec.scala
Comment (vkostyukov) [inline core/src/test/scala/io/finch/EndToEndSpec.scala:39]: Don't forget to remove this.

File: core/src/test/scala/io/finch/EndpointSpec.scala
Comment (vkostyukov) [inline core/src/test/scala/io/finch/EndpointSpec.scala:451]: Same: there might be already an instance of `Async` for Writer so we can summon Dispatcher.

File: core/src/test/scala/io/finch/MissingInstances.scala
Comment (vkostyukov) [inline core/src/test/scala/io/finch/MissingInstances.scala:15]: We probably wouldn't need to do that if we defer to `Dispatcher[IO].use` API.

File: examples/src/main/scala/io/finch/middleware/Main.scala
Comment (vkostyukov) [inline examples/src/main/scala/io/finch/middleware/Main.scala:23]: Oh that's very cool. We should do that for other examples too if we can.

File: examples/src/main/scala/io/finch/wrk/Finch.scala
Comment (vkostyukov) [inline examples/src/main/scala/io/finch/wrk/Finch.scala:25]: We wouldn't need to do that if stop requiring an implicit Dispatcher. It would be easier to interact with Endpoint API that way.

**Review by vkostyukov — CHANGES_REQUESTED** [review]
This is very cool, @tpetillot!

I'm totally onboard to drop CE2 and move to CE3, but I think we'd need to figure out the proper use of a `Dispatcher` utility, which feels a little out of order now.

File: benchmarks/src/main/scala/io/finch/benchmarks.scala
Comment (joroKr21) [reply benchmarks/src/main/scala/io/finch/benchmarks.scala:26]: Yes there is but it returns a `Resource`

File: core/src/main/scala/io/finch/Endpoint.scala
Comment (joroKr21) [reply core/src/main/scala/io/finch/Endpoint.scala:696]: `Async` already has a blocking EC baked in

File: core/src/main/scala/io/finch/EndpointResult.scala
Comment (joroKr21) [reply core/src/main/scala/io/finch/EndpointResult.scala:43]: Wouldn't that change the signature?

File: core/src/main/scala/io/finch/Bootstrap.scala
Comment (tpetillot) [reply core/src/main/scala/io/finch/Bootstrap.scala:85]: I'm not very confident about this... Looking at `Dispatcher[IO]`, it seems to me that instantiation is expensive, but at the same time doesn't seem to be that much.

Looking forward, I've sturborned across a discussion between [Fabio Labella and @joroKr21 on Gitter ](https://gitter.im/typelevel/cats-effect?at=5fbebdbafe857d71dc953fe3)on the same concern, where it's stated that Dispatcher instantiation:

> isn't that heavyweight either, just a Fiber and some AtomicReferences
> it's very very lightweight in fact

Comment (tpetillot) [reply core/src/main/scala/io/finch/Bootstrap.scala:85]: But no matter, dispatcher instanciation return a `Resource[F, Dispatcher[F]]`, which we cannot run synchronously on its own...

File: core/src/main/scala/io/finch/Endpoint.scala
Comment (tpetillot) [reply core/src/main/scala/io/finch/Endpoint.scala:696]: @joroKr21 thanks, completly missed it!

File: build.sbt
Comment (tpetillot) [reply build.sbt:19]: Agree, removed!

File: core/src/main/scala/io/finch/Bootstrap.scala
Comment (vkostyukov) [reply core/src/main/scala/io/finch/Bootstrap.scala:85]: I guess that's my point exactly? Shouldn't we `use` that resource instead of accessing Dispatcher directly? I think we're side-stepping the API for some reason.

This is the use method I'm referring to: https://github.com/typelevel/cats-effect/blob/series/3.x/kernel/shared/src/main/scala/cats/effect/kernel/Resource.scala#L206

Comment (joroKr21) [reply core/src/main/scala/io/finch/Bootstrap.scala:85]: `use` also would change the signature to `IO`

Comment (vkostyukov) [reply core/src/main/scala/io/finch/Bootstrap.scala:85]: Ok, I see. I'm reading this doc (https://typelevel.org/cats-effect/docs/migration-guide) and it, at least, says that we should use Dispatcher as a type-class, b/c it's not one:

> An alternative approach to creating a Dispatcher on our own would be taking one as a parameter. It is recommended to pass it explicitly, as Dispatcher isn't a type class and there can be many instances of it in an application.

We somehow need to bubble Dispatcher construction up, I assume. Should it maybe happen as part of a Bootstrap? Like Bootstrap.toService would return you a `Resource[Service]` with `Dispatcher` fused in?

I understand we also need to figure out how to run those EndpointResult methods since they are happening outside of Bootstrap. Maybe for those we should use ask for Dispatcher explicitly and expect users to instantiate one themselves? Maybe it's not too much trouble since it's only happening in tests.

Comment (tpetillot) [reply core/src/main/scala/io/finch/Bootstrap.scala]: Updated accordingly

File: core/src/test/scala/io/finch/BodySpec.scala
Comment (vkostyukov) [inline core/src/test/scala/io/finch/BodySpec.scala:39]: Maybe CE already provides some kind of test base class that instantiates a dispatcher on before and shuts it down on after?

File: examples/src/main/scala/io/finch/iteratee/Main.scala
Comment (vkostyukov) [inline examples/src/main/scala/io/finch/iteratee/Main.scala:76]: This structure makes so much sense now, I love it!

File: examples/src/main/scala/io/finch/middleware/Main.scala
Comment (vkostyukov) [inline examples/src/main/scala/io/finch/middleware/Main.scala:57]: An idea for another PR: maybe we should make this function a utility and provide our wrapper around IOApp.

File: examples/src/main/scala/io/finch/todo/Main.scala
Comment (vkostyukov) [inline examples/src/main/scala/io/finch/todo/Main.scala:54]: Can we do this instead?

```
val run = Dispatcher[IO].use { dispathcer =>
  // run on the things
}

// since run is IO
run.unsafeRunSync()
```

**Review by vkostyukov — COMMENTED** [review]
This looks pretty awesome! Thank you for following through the dispatcher changes, @tpetillot!

I think there is a few places left that we might need to clean up, but we're almost there. I want to look around the CE3 APIs today to see if we're missing anything else. I like the use of Resouce[F, Service[_, _]] already but perhaps we'd need to spend a little more thinking to make sure we don't compromise the testing experience for users.

A couple of things I want to find answers to:
 - In tests (our own and users), can we somehow avoid instantiating our own dispatcher
 - Can we make experience of EndpointResult.await methods better?

File: examples/src/main/scala/io/finch/iteratee/Main.scala
Comment (joroKr21) [inline examples/src/main/scala/io/finch/iteratee/Main.scala:77]:
```suggestion
    } yield server).useForever
```

File: examples/src/main/scala/io/finch/middleware/Main.scala
Comment (joroKr21) [inline examples/src/main/scala/io/finch/middleware/Main.scala:65]:
```suggestion
    } yield server).useForever
```

File: examples/src/main/scala/io/finch/wrk/Finagle.scala
Comment (joroKr21) [inline examples/src/main/scala/io/finch/wrk/Finagle.scala:13]: This line is a duplicate

File: examples/src/main/scala/io/finch/wrk/Finch.scala
Comment (joroKr21) [inline examples/src/main/scala/io/finch/wrk/Finch.scala:15]: Delete this line

File: examples/src/main/scala/io/finch/wrk/Finch.scala
Comment (joroKr21) [inline examples/src/main/scala/io/finch/wrk/Finch.scala:37]:
```suggestion
    } yield server).useForever
```

File: examples/src/main/scala/io/finch/wrk/Finch.scala
Comment (joroKr21) [inline examples/src/main/scala/io/finch/wrk/Finch.scala:26]: Why do we need it? If we use `IOApp`?

**Review by joroKr21 — CHANGES_REQUESTED** [review]
This looks much better than I thought it would.
See my comments for some improvements, but overall amazing work 👍

File: examples/src/main/scala/io/finch/middleware/Main.scala
Comment (vkostyukov) [reply examples/src/main/scala/io/finch/middleware/Main.scala:65]: Wow, sweet!

File: core/src/test/scala/io/finch/BodySpec.scala
Comment (vkostyukov) [reply core/src/test/scala/io/finch/BodySpec.scala:39]: Thinking out loud: what if we give up providing this awaitX method on EndpointResult and maybe move that responsibility to users? If you think about it, it's not too bad of an experience after all, given that F is already defined as IO. Folks can just can runSyncUnsafe  (or whatever it's called) on their IO instances in tests directly. We could still make this experience nicer if we provide convenient functions transforming the output, but running it should be something we leave our users to deal with.

Comment (joroKr21) [reply core/src/test/scala/io/finch/BodySpec.scala:39]: That's surely an option - but we would have to change our own tests quite a bit. I would leave that for another PR.

File: core/src/test/scala/io/finch/FinchSpec.scala
Comment (tpetillot) [reply core/src/test/scala/io/finch/FinchSpec.scala:292]: Dispatcher use promise here: https://github.com/typelevel/cats-effect/blob/series/3.x/std/shared/src/main/scala/cats/effect/std/Dispatcher.scala#L194
And promise box error here: https://github.com/scala/scala/blob/2.12.x/src/library/scala/concurrent/impl/Promise.scala#L87

Breaking test when `Arbitrary.arbitrary[Throwable]` throw an `Error`.

File: core/src/main/scala/io/finch/endpoint/endpoint.scala
Comment (tpetillot) [reply core/src/main/scala/io/finch/endpoint/endpoint.scala:21]: Nice catch! Added defer for stack safety.

File: core/src/test/scala/io/finch/internal/ToEffectSpec.scala
Comment (tpetillot) [reply core/src/test/scala/io/finch/internal/ToEffectSpec.scala:14]: Used by `unsafeRunSync()`, for a reason I don't understand, when importing `cats.effect.unsafe.implicits.global` implicit is not resolved. 😞

File: examples/src/main/scala/io/finch/middleware/Main.scala
Comment (tpetillot) [reply examples/src/main/scala/io/finch/middleware/Main.scala:57]: Definitely! Is there any place to add feature request?

File: core/src/main/scala/io/finch/Input.scala
Comment (tpetillot) [reply core/src/main/scala/io/finch/Input.scala]: Updated to take it explicitly

File: benchmarks/src/main/scala/io/finch/benchmarks.scala
Comment (vkostyukov) [inline benchmarks/src/main/scala/io/finch/benchmarks.scala:204]: I think it's ok if we keep this as is for this change, but we'd need to revisit this sometime later. Ideally, we'd need to `use` the resource on each benchmark invocation. Maybe we could leverage before/after setup machine to manage the lifecycle of the resource.

File: core/src/test/scala/io/finch/BodySpec.scala
Comment (vkostyukov) [reply core/src/test/scala/io/finch/BodySpec.scala:39]: Yea I like your plan.

File: core/src/main/scala/io/finch/Input.scala
Comment (vkostyukov) [inline core/src/main/scala/io/finch/Input.scala:106]: This is another place that I'm not in love with but we should leave it for another PR. I don't have a ton of ideas just yet but I'm thinking to extract this and awaitX methods into their own testing package (or syntax?) that's hopefully structured around the existing dispatcher.

File: core/src/test/scala/io/finch/FinchSpec.scala
Comment (vkostyukov) [inline core/src/test/scala/io/finch/FinchSpec.scala:309]: Oh this eq methods need more attention. Maybe we could move these under missingInstances that's instantiated around the existing dispatcher.

File: examples/src/main/scala/io/finch/middleware/Main.scala
Comment (vkostyukov) [reply examples/src/main/scala/io/finch/middleware/Main.scala:57]: Maybe just file a Github issue for now?

File: examples/src/main/scala/io/finch/todo/Main.scala
Comment (vkostyukov) [inline examples/src/main/scala/io/finch/todo/Main.scala:50]: Should we drop IO.never here? I think it we just keep waiting on `Await.ready`, the app would correctly shutdown when /quitquitqui endpoint is hit.

**Review by vkostyukov — APPROVED** [review]
This looks pretty nice @tpetillot!

I think I'm ok merging this PR in the current form but keep iterating on the testing pieces. We'd probably want to be good citizens of the CE3 API and stop instantiating our own dispatchers.

I'd wait a ship it from @joroKr21 before merging. And please, use squash commits when merging so we keep the git history more or less clean.

**Review by joroKr21 — APPROVED** [review]
I think we can merge and think a bit more about dispatcher and testing.

File: core/src/test/scala/io/finch/FinchSpec.scala
Comment (joroKr21) [reply core/src/test/scala/io/finch/FinchSpec.scala:292]: Ok weird, TIL

File: core/src/test/scala/io/finch/FinchSpec.scala
Comment (tpetillot) [reply core/src/test/scala/io/finch/FinchSpec.scala:309]: Existing dispatcher is already set on IO (and eq is used based on higher kind). I'll let it that way until next PR 🚀

**Comment by tpetillot** [comment]
@joroKr21 @vkostyukov merged 🎊 
Looking forward next propositions regarding dispatching (I will try to look into it a bit more as well).

In anticipation of our future upgrades to CE3, when would you estimate a first release candidate?

**Comment by vkostyukov** [comment]
I was thinking we do a couple more changes:
 - testing and dispatcher
 - #1471 
 
 And do a RC1 for 1.0. What do you both think about this plan?

**Comment by joroKr21** [comment]
> And do a RC1 for 1.0. What do you both think about this plan?

Sounds good 👍

**Pull request merged**
